src/os/linux/lwpa/os_netint.c

```diff
@@ -17,4 +17,682 @@
+ * RtNetlink sockets: http://man7.org/linux/man-pages/man7/rtnetlink.7.html
+ * Some sample RtNetlink code: https://www.linuxjournal.com/article/8498
+ */
+
 #include "lwpa/netint.h"
+
+#include <stdlib.h>
+#include <limits.h>
+#include <string.h>
+
+#ifdef LWPA_NETINT_DEBUG_OUTPUT
+#include <stdio.h>
+#endif
+
+#include <ifaddrs.h>
+#include <errno.h>
+#include <asm/types.h>
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "lwpa/socket.h"
+#include "lwpa/private/netint.h"
+#include "os_error.h"
+
+/***************************** Private types *********************************/
+
+typedef struct RoutingTableEntry
+{
+  LwpaIpAddr addr;
+  LwpaIpAddr mask;
+  LwpaIpAddr gateway;
+  int interface_index;
+  int metric;
+} RoutingTableEntry;
+
+typedef struct RoutingTable
+{
+  RoutingTableEntry* entries;
+  RoutingTableEntry* default_route;
+  size_t size;
+} RoutingTable;
+
+/* A composite struct representing an RT_NETLINK request sent over a netlink socket. */
+typedef struct RtNetlinkRequest
+{
+  struct nlmsghdr nl_header;
+  struct rtmsg rt_msg;
+} RtNetlinkRequest;
+
+/**************************** Private variables ******************************/
+
+static struct LwpaNetintState
+{
+  bool initialized;
+
+  RoutingTable routing_table_v4;
+  RoutingTable routing_table_v6;
+
+  struct ifaddrs* os_addrs;
+  size_t num_netints;
+  LwpaNetintInfo* lwpa_netints;
+
+  bool have_default_netint_index_v4;
+  size_t default_netint_index_v4;
+
+  bool have_default_netint_index_v6;
+  size_t default_netint_index_v6;
+} state;
+
+/*********************** Private function prototypes *************************/
+
+// Functions for building the routing tables
+static lwpa_error_t build_routing_tables();
+static lwpa_error_t build_routing_table(int family, RoutingTable* table);
+static void free_routing_tables();
+static void free_routing_table(RoutingTable* table);
+
+// Interacting with RTNETLINK
+static lwpa_error_t send_netlink_route_request(int socket, int family);
+static lwpa_error_t receive_netlink_route_reply(int sock, int family, size_t buf_size, RoutingTable* table);
+static lwpa_error_t parse_netlink_route_reply(int family, const char* buffer, size_t nl_msg_size, RoutingTable* table);
+
+// Manipulating routing table entries
+static void init_routing_table_entry(RoutingTableEntry* entry);
+static int compare_routing_table_entries(const void* a, const void* b);
+
+// Functions for enumerating the interfaces
+static lwpa_error_t enumerate_netints();
+static int compare_netints(const void* a, const void* b);
+static void free_netints();
+
+#if LWPA_NETINT_DEBUG_OUTPUT
+static void debug_print_routing_table(RoutingTable* table);
+#endif
+
+/*************************** Function definitions ****************************/
+
+lwpa_error_t lwpa_netint_init()
```

Comment:
I feel like there's some overlap between this and windows.  Could there be some shared code to get rid of boilerplate?  Or is there enough differences that it isn't worth it..

Reply:
I'll take a second look. Last time I looked at it it was a bit too different.